basic.go

```diff
@@ -93,67 +93,83 @@ func (b *basicBus) tryDropNode(typ reflect.Type) {
 	b.lk.Unlock()
 }
 
+type sub struct {
+	ch      chan interface{}
+	nodes   []*node
+	dropper func(reflect.Type)
+}
+
+func (s *sub) Out() <-chan interface{} {
+	return s.ch
+}
+
+func (s *sub) Close() error {
+	close(s.ch)
+	for _, n := range s.nodes {
+		n.lk.Lock()
+		for i := 0; i < len(n.sinks); i++ {
+			if n.sinks[i] == s.ch {
+				n.sinks[i], n.sinks[len(n.sinks)-1] = n.sinks[len(n.sinks)-1], nil
+				n.sinks = n.sinks[:len(n.sinks)-1]
+				break
+			}
+		}
+		tryDrop := len(n.sinks) == 0 && atomic.LoadInt32(&n.nEmitters) == 0
+		n.lk.Unlock()
+		if tryDrop {
+			s.dropper(n.typ)
+		}
+	}
+	return nil
+}
+
+var _ event.Subscription = (*sub)(nil)
+
 // Subscribe creates new subscription. Failing to drain the channel will cause
 // publishers to get blocked. CancelFunc is guaranteed to return after last send
 // to the channel
-//
-// Example:
-// ch := make(chan EventT, 10)
-// defer close(ch)
-// cancel, err := eventbus.Subscribe(ch)
-// defer cancel()
-func (b *basicBus) Subscribe(typedChan interface{}, opts ...event.SubscriptionOpt) (c event.CancelFunc, err error) {
+func (b *basicBus) Subscribe(evtTypes interface{}, opts ...event.SubscriptionOpt) (_ event.Subscription, err error) {
 	var settings subSettings
 	for _, opt := range opts {
 		if err := opt(&settings); err != nil {
 			return nil, err
 		}
 	}
 
-	refCh := reflect.ValueOf(typedChan)
-	typ := refCh.Type()
-	if typ.Kind() != reflect.Chan {
-		return nil, errors.New("expected a channel")
+	types, ok := evtTypes.([]interface{})
+	if !ok {
+		types = []interface{}{evtTypes}
 	}
-	if typ.ChanDir()&reflect.SendDir == 0 {
-		return nil, errors.New("channel doesn't allow send")
+
+	out := &sub{
+		ch:    make(chan interface{}, settings.buffer),
+		nodes: make([]*node, len(types)),
+
+		dropper: b.tryDropNode,
 	}
 
-	if settings.forcedType != nil {
-		if settings.forcedType.Elem().AssignableTo(typ) {
-			return nil, fmt.Errorf("forced type %s cannot be sent to chan %s", settings.forcedType, typ)
+	for i, etyp := range types {
+		typ := reflect.TypeOf(etyp)
+
+		if typ.Kind() != reflect.Ptr {
+			return nil, errors.New("subscribe called with non-pointer type")
```

Comment:
This needs to close already created subscriptions.

Reply:
👍 yeah, since this is merged and released, could you open an issue for it?